crates/server/src/lib.rs

```diff
@@ -183,32 +192,69 @@ impl Server {
 
         let server = axum::Server::from_tcp(listener)?.serve(
             Self::create_router(
-                format!("http://{addr}", addr = self.config.addr.unwrap()),
-                self.config.content_dir,
+                format!("http://{addr}", addr = endpoints.api),
+                self.config.content_dir.take(),
                 core,
             )?
             .into_make_service(),
         );
 
-        tracing::info!("listening on {addr}", addr = self.config.addr.unwrap());
+        tracing::info!("api endpoint listening on {addr}", addr = endpoints.api);
 
-        if let Some(shutdown) = self.config.shutdown {
-            tracing::debug!("server is running with a shutdown signal");
+        // Shut down core service when token cancelled.
+        let token = self.token.clone();
+        self.tasks.spawn(async move {
+            token.cancelled().await;
+            tracing::info!("waiting for core service to stop");
+            handle.stop().await;
+            Ok(())
+        });
```

Comment:
Could this task be merged with the next task? That is to say, after the `server` future completes in the next task, we simply wait on the stop handle for the core service.

Given this design, I would expect a 1:1 between a task in the task set and a running server (API in this case, but soon a monitoring server, etc). Each would shutdown in their own way.

Reply:
I'd think due to the nature of fanning out a cancelation signal chain, a cancellation token should be cloned for each owned depedency, somewhat similar to how [Go Context](https://pkg.go.dev/context) works. Otherwise there virtually will be for-loop boilerplate being written to cancel "children" all separately as a server grows. That allows things to shut down in parallel easily if needed.

But yes, could keep things simple for now and merge the shutdown tasks instead of renaming chaining of them through the tokens themselves. However I'd either want to use a broadcast channel instead, or an additional token for draining in #111.

Reply:
I don't think we want the core service to be stopped in parallel to the hyper server shutting down as it might race.

If we call `stop` on the core service, it will break out of its loop and the task will complete; note that the `CoreService` instance itself won't be dropped as a result of this as there are references to it from the hyper server.

If this is done in parallel to the hyper server shutting down, then it's possible a request being drained attempts to enqueue work for the core service to do; if it so happens that the core service already stopped and the send limit for the message queue is reached, it'll block waiting for the core service to remove a message which will never occur since its task has ended.

We also don't want to close the message queue's sender as part of stopping the core service as this would result in a panic (even a graceful error back to the client here is not ideal) and a reset of the connection for any request attempting to enqueue work for the core service to do after it has stopped.

I believe the correct order of operations here is to shutdown the hyper server then join on the core service, like we had before.

Reply:
Note that eventually the core service won't be part `warg-server`, but instead another backend service for it to talk to to process new record entries; it's the limiting factor of why we can't spin up multiple `warg-server` instances right now.